.github/workflows/tests.yaml

```diff
@@ -0,0 +1,160 @@
+name: Tests
+
+on:
+  push:
+  pull_request:
+
+concurrency:
+  group: tests-${{ github.head_ref || github.ref }}
+  cancel-in-progress: ${{ github.event_name == 'pull_request' }}
+
+defaults:
+  run:
+    shell: bash
+
+jobs:
+  changes:
+    name: Detect changed files
+    runs-on: ubuntu-latest
+    permissions:
+      pull-requests: read
+    outputs:
+      fixtures-pypi: ${{ steps.changes.outputs.fixtures-pypi }}
+      pytest: ${{ steps.changes.outputs.pytest }}
+    steps:
+      - uses: actions/checkout@v4
+      - uses: dorny/paths-filter@v3
+        id: changes
+        with:
+          filters: |
+            workflow: &workflow
+              - '.github/workflwos/tests.yaml'
+            fixtures-pypi:
+              - *workflow
+              - 'tests/repositories/fixtures/pypi.org/**'
+            pytest:
+              - *workflow
+              - 'poetry.lock'
+              - 'pyproject.toml'
+              - 'src/**.py'
+              - 'tests/**'
+
+  fixtures-pypi:
+    name: PyPI Fixtures
+    runs-on: ubuntu-latest
+    needs: changes
+    if: needs.changes.outputs.fixtures-pypi == 'true'
+    steps:
+      - uses: actions/checkout@v4
+
+      - name: Bootstrap poetry
+        run: pipx install poetry
+
+      - name: Set up Python
+        uses: actions/setup-python@v5
+        with:
+          python-version: "3.12"
+          cache: poetry
+
+      - name: Configure poetry
+        run: |
+          poetry config virtualenvs.in-project true
+          poetry config virtualenvs.options.no-pip true
+          poetry config virtualenvs.options.no-setuptools true
+
+      - name: Install dependencies
+        run: poetry install --only main,test
+
+      - name: Regenerate PyPI fixtures
+        run: PYTHONPATH="$PWD" poetry run python tests/repositories/fixtures/pypi.org/generate.py
+
+      - name: Check for changed files
+        run: git diff --exit-code --stat HEAD tests/repositories/fixtures/pypi.org
+
+  pytest:
+    name: ${{ matrix.os }} / ${{ matrix.python-version }}
+    runs-on: ${{ matrix.image }}
+    needs: changes
+    strategy:
+      matrix:
+        os: [Ubuntu, macOS, Windows]
+        python-version: ["3.8", "3.9", "3.10", "3.11", "3.12"]
+        include:
+          - os: Ubuntu
+            image: ubuntu-22.04
+          - os: Windows
+            image: windows-2022
+          - os: macOS
+            image: macos-12
+      fail-fast: false
+    steps:
+      - uses: actions/checkout@v4
+        if: needs.changes.outputs.pytest == 'true'
+
+      - name: Bootstrap poetry
+        if: needs.changes.outputs.pytest == 'true'
```

Comment:
Can't we do this at the job level? Should still work with required checks.

Reply:
It won't, because we won't expand the matrix. I am working on solutions to that for the 'DRY things up' follow-up I have in mind (likely a single job that checks all the other jobs, keeping us from having to adjust the required checks).